### PR TITLE
Yaml2ck script

### PR DESCRIPTION
**Pull request by bryanwweber**
**Changes proposed in this pull request**

- Add a `yaml2ck` script to convert YAML input files/Solution instances to Chemkin file(s)
- ~TODO: Add a method to `Solution` to import and call `yaml2ck.convert`~
- ~TODO: Fix up the docstring~

**If applicable, fill in the issue number this pull request is fixing**

Supersedes #1185 
I tried to push an update to that branch and it closed the PR 🤷‍♂️ 

This also relies on the work in #1266 for some of the sorting, so it should be merged after that PR.

**If applicable, provide an example illustrating new features this pull request is introducing**

```python
from cantera import yaml2ck
yaml2ck.convert("input_file.yaml")
```

**Checklist**

- [X] The pull request includes a clear description of this code change
- [X] Commit messages have short titles and reference relevant issues
- [X] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [X] Style & formatting of contributed code follows [contributing guidelines](https://github.com/Cantera/cantera/blob/main/CONTRIBUTING.md)
- [X] The pull request is ready for review

### CODE REVIEW

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py]:
```suggestion
            fmt_data["note"] = ""

        temperature_range = input_data["thermo"]["temperature-ranges"]
```
Not sure what *black* thinks about this, but I find it helpful to have a blank line after an `if-else` block. There is another instance below.

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py:339]: I believe there was some discussion about the intermediate temperature defaulting to 1000? Is this a safe assumption? It would be good to document ...

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py:606]: While `None` is less clear as user output, it's more pythonic. Certainly worth considering *elsewhere* :smile:

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py]:
```suggestion
    parser.add_argument(
        "--mechanism", help="The output mechanism filename.", metavar="MECH"
    )
```
Having a shortened `metavar` will produce nicer output for `yaml2ck --help`.

As an aside, I believe there is also an option where this argument changes the output path?

PS: renaming `--mechanism` to `--output` would be an option also (it would be intuitive from the CLI).

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py]:
```suggestion
    parser.add_argument(
        "--phase-name", 
        help=("Name of the phase to load from the input. Optional argument if input "
              "file contains more than one phase definition. Cantera's default "
              "behavior will load the first definition from the file."),
        metavar="PHASE"
    )
```
Or similar text clarifying what this does.

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py]:
```suggestion
    parser.add_argument(
        "--transport", help="The output transport filename.", metavar="TRAN"
    )
```
Probably need a description of what is done if it is omitted (similar for `--thermo`). It may make sense to copy some of the more comprehensive docstrings over from `convert` as I anticipate that many users may use `yaml2ck` from the CLI.

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py]:
```suggestion
    parser.add_argument(
        "--overwrite",
        action="store_true",
        default=False,
        help="If set, overwrite existing output files.",
    )
```
Same for `--validate` below.

Together with the other suggestions for `metavar` (and moving `--overwrite` and `--validate` up, the formatting of `yaml2ck --help` becomes a lot easier to parse visually, i.e.
```
(cantera-dev) PS C:\Users\ischo\GitHub\cantera> yaml2ck --help
usage: yaml2ck [-h] [--phase-name PHASE] [--mechanism MECH] [--thermo THERMO] [--transport TRAN] [--overwrite]
               [--validate] [--sort-elements {None,alphabetical,molecular-weight}]
               [--sort-species {None,alphabetical,molecular-weight}]
               input

Convert Cantera YAML input files to Chemkin-format mechanisms

positional arguments:
  input                 The input YAML filename. Required.

options:
  -h, --help            show this help message and exit
  --phase-name PHASE    Name of the phase to load from the input. Optional argument if input file contains more than
                        one phase definition. Cantera's default behavior will load the first definition from the file.
  --mechanism MECH      The output mechanism filename.
  --thermo THERMO       The output thermodynamics filename.
  --transport TRAN      The output transport filename.
  --overwrite           If set, overwrite existing output files.
  --validate            If set, check that the mechanism can be loaded back into Cantera.
  --sort-elements {None,alphabetical,molecular-weight}
                        Sort elements in source order (None), alphabetically, or from lowest to highest atomic mass.
  --sort-species {None,alphabetical,molecular-weight}
                        Sort species in source order (None), alphabetically, or from lowest to highest molecular
                        weight.
```

There are additional formatting options [documented on SO](https://stackoverflow.com/questions/52605094/python-argparse-increase-space-between-parameter-and-description)

**Review by ischoegl — COMMENTED** [review]
@bryanwweber ... thanks for picking this up! This overall looks really good, and I don't have any significant comments (other than formatting suggestions for `yaml2ck --help`).

Running this on my (windows) machine, I am, however, running into the following issue:
```
(cantera-dev) PS C:\Users\ischo\GitHub\cantera> python --version
Python 3.10.2
(cantera-dev) PS C:\Users\ischo\GitHub\cantera> yaml2ck .\h2o2.yaml
Traceback (most recent call last):
  File "C:\Users\ischo\miniconda3\envs\cantera-dev\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\ischo\miniconda3\envs\cantera-dev\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\ischo\miniconda3\envs\cantera-dev\Scripts\yaml2ck.exe\__main__.py", line 7, in <module>
  File "C:\Users\ischo\miniconda3\envs\cantera-dev\lib\site-packages\cantera\yaml2ck.py", line 685, in main
    output_paths = convert(
  File "C:\Users\ischo\miniconda3\envs\cantera-dev\lib\site-packages\cantera\yaml2ck.py", line 511, in convert
    solution = ct.Solution(solution, phase_name)
  File "build\python\cantera\base.pyx", line 71, in cantera._cantera._SolutionBase.__cinit__
    self._cinit(infile=infile, name=name, adjacent=adjacent, origin=origin,
  File "build\python\cantera\base.pyx", line 126, in cantera._cantera._SolutionBase._cinit
    self._init_yaml(infile, name, adjacent, yaml, transport_model)
  File "build\python\cantera\base.pyx", line 210, in cantera._cantera._SolutionBase._init_yaml
    stringify(infile), stringify(name), cxx_transport, adjacent_solns)
  File "build\python\cantera\utils.pyx", line 27, in cantera._cantera.stringify
    tmp = bytes(x.encode())
AttributeError: 'NoneType' object has no attribute 'encode'
```
despite `scons test-python-convert` passing.

File: .github/workflows/main.yml
Comment (ischoegl) [inline .github/workflows/main.yml:415]: It may be worth adding 
```
yaml2ck test/data/h2o2.yaml --mechanism=h2o2-test.ck 
```
in the `Test Essentials` entry for this runner? (same for the `windows-2022` runner)

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py]:
```suggestion
        choices=[None, "alphabetical", “weight"],
```
Shorten to `weight`, as it doesn’t sound right for elements. For symmetry, also shorten for species?

Comment (bryanwweber) [reply interfaces/cython/cantera/yaml2ck.py]: What about `molar-mass`? For both?

File: interfaces/cython/cantera/yaml2ck.py
Comment (bryanwweber) [reply interfaces/cython/cantera/yaml2ck.py:606]: I suspect it's the `None` string to avoid the problem you ran into on Windows, but it's probably possible to be more flexible.

File: interfaces/cython/cantera/yaml2ck.py
Comment (bryanwweber) [reply interfaces/cython/cantera/yaml2ck.py:339]: The thread is here: https://github.com/Cantera/cantera/pull/1185#discussion_r828649674 The problem is that CHEMKIN kinda sorta uses the McBride NASA format, but also has some changes that mean the values here are never actually used for anything, since each species is required to have its own definition of all three temperatures according to CHEMKIN (general NASA programs do not require that, apparently?). Anyhow, I'm not sure what to say here that would be coherent and helpful for a future reader... I'll try to add something useful 😄

**Comment by bryanwweber** [comment]
@ischoegl I think I fixed the problem you posted about `NoneType`, can you try now?

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [reply interfaces/cython/cantera/yaml2ck.py:339]: That would be great!

**Comment by ischoegl** [comment]
@bryanwweber ... better, but not quite:
```
(cantera-dev) PS C:\Users\ischo\GitHub\cantera> yaml2ck data/h2o2.yaml --mechanism=h2o2-test.ck --overwrite
Validating mechanism...PASSED.
Traceback (most recent call last):
  File "C:\Users\ischo\miniconda3\envs\cantera-dev\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\ischo\miniconda3\envs\cantera-dev\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\ischo\miniconda3\envs\cantera-dev\Scripts\yaml2ck.exe\__main__.py", line 7, in <module>
  File "C:\Users\ischo\miniconda3\envs\cantera-dev\lib\site-packages\cantera\yaml2ck.py", line 758, in main
    output = "\n".join(ck_paths)
TypeError: sequence item 0: expected str instance, WindowsPath found
```

File: .github/workflows/main.yml
Comment (ischoegl) [inline .github/workflows/main.yml]:
```suggestion
          yaml2ck data/h2o2.yaml --mechanism=h2o2-test.ck
```
Same for other instance

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py]:
```suggestion
                    f"Output file '{fil}' exists. Remove it or specify '--overwrite'"
```

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py]:
```suggestion
        "--no-validate",
        action="store_true",
        default=False,
```
Otherwise validate would be run in all cases?

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py:751]: I believe `--output` would be more intuitive here.

**Review by ischoegl — COMMENTED** [review]
@bryanwweber .... Thanks! Again, pretty minor things.

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [inline interfaces/cython/cantera/yaml2ck.py]:
```suggestion
    output = "\n".join([str(p) for p in ck_paths])
    print(f"Output written to: \n{output}")
```
With that, things run through on my machine!

**Comment by ischoegl** [comment]
Before I forget ... `yaml2ck` this should probably marked with `.. versionadded:: 3.0`

File: interfaces/cython/cantera/yaml2ck.py
Comment (bryanwweber) [reply interfaces/cython/cantera/yaml2ck.py]: Ah, with the way this was before, you could do `--validate=False` to turn off validation, but I think it probably makes sense to have `--validate` and `--no-validate`

Comment (ischoegl) [reply interfaces/cython/cantera/yaml2ck.py]: if validation is the default, then `--validate` isn't needed?

Comment (bryanwweber) [reply interfaces/cython/cantera/yaml2ck.py]: There's a nice action (present in 3.9+) that automatically handles this kind of thing. I'll add that. I think this kind of behavior is typical for optional Boolean arguments.

File: interfaces/cython/cantera/yaml2ck.py
Comment (bryanwweber) [reply interfaces/cython/cantera/yaml2ck.py:751]: `--mechanism` is I think the typical name for the file, plus with (possibly) three separate file outputs, I don't think that `--output` would capture that. I think `--output` could make sense as the folder output option, but I'm concerned about adding too many options...

File: 1
Comment (ischoegl) [inline 1]: Not sure what this file is?

**Review by ischoegl — COMMENTED** [review]
Everything else looks good

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [reply interfaces/cython/cantera/yaml2ck.py:751]: Not a big issue …

File: interfaces/cython/cantera/yaml2ck.py
Comment (ischoegl) [reply interfaces/cython/cantera/yaml2ck.py:339]: Could you add this comment?

File: 1
Comment (bryanwweber) [reply 1]: This is a result of me _always_ getting the shell redirection syntax backwards... `python -m http.server 2&>1` should be `2>&1` to redirect to `stdout` :joy:

File: interfaces/cython/cantera/yaml2ck.py
Comment (bryanwweber) [reply interfaces/cython/cantera/yaml2ck.py:339]: Thanks for the reminder, done now.

**Comment by ischoegl** [comment]
@bryanwweber … will approve this once the merge conflict is resolved

**Review by ischoegl — APPROVED** [review]
LGTM :tada:

**Comment by bryanwweber** [comment]
Tagging @tsikes thanks for all your work here!

**Comment by tsikes** [comment]
@bryanwweber Thank you so much for carrying the torch the rest of the way! I'm happy to see this completed.

**Pull request merged**
